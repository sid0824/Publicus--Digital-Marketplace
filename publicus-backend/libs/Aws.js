import AWS from 'aws-sdk';
import {AccessKeyId, SecretAccessKey} from "../config/ConfigValues";

export const s3 = new AWS.S3({
  region: 'ap-southeast-2',
  accessKeyId: AccessKeyId,
  secretAccessKey: SecretAccessKey
});

export const ses = new AWS.SES({
  region: 'us-west-2'
});