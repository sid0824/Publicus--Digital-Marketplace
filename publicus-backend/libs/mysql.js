import mysql from 'mysql';
import {MySqlHost, MySqlPassword, MySqlUser} from "../config/ConfigValues";

//mysql connection to database
const connection = mysql.createConnection({
    host: MySqlHost,
    user: MySqlUser,
    password: MySqlPassword,
    database: "survey",
    multipleStatements: true
});

export const runQuery = async (query) => {
    return new Promise((resolve, reject) => {
        connection.query(query, (error, results) => {
            if(error) {
                reject(error);
            } else {
                resolve(results);
            }
        });
    });
};