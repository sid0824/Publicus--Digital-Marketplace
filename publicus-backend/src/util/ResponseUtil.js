export function success(body) {
    return buildResponse(200, body);
}

export function failure(body) {
    return buildResponse(500, body);
}

export function json_success(body) {
    return jsonResponse(200, body);
}

export function unauthorised(body) {
    return buildResponse(401, body);
}

function buildResponse(statusCode, body) {
    return {
        statusCode: statusCode,
        headers: {
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Credentials": true,
        },
        body: JSON.stringify(body)
    };
}

function jsonResponse(statusCode, body) {
    return {
        statusCode: statusCode,
        headers: {
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Credentials": true,
            "Content-Type": "applications/json"
        },
        body: JSON.stringify(body)
    };
}
