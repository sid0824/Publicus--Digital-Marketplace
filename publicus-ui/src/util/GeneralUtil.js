/**
 * Capitalises the first letter of the provided string
 */
export const capitaliseFirstLetter = (string) => string.charAt(0).toUpperCase() + string.slice(1);

