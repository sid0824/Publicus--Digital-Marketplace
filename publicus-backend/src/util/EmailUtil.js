import { ses } from '../../libs/Aws';
import { SiteUrl, SourceEmail } from "../../config/ConfigValues";

const sendEmail = async (email, subject, body) => {
  const params = {
    Destination: {
      ToAddresses: [
        email
      ]
    },
    Message: {
      Body: {
        Text: {
          Charset: "UTF-8",
          Data: body
        }
      },
      Subject: {
        Charset: "UTF-8",
        Data: subject
      }
    },
    Source: SourceEmail,
  };
  try {
    await ses.sendEmail(params).promise();
  } catch (err) {
    console.error(err);
  }
};

export const sendPurchasedEmail = async (email) => {
  const body = `One of your forms was purchased on the Publicus Marketplace. View in the site.\n${SiteUrl}/myForms`;
  const subject  = "A User Purchased Your Form";
  await sendEmail(email, subject, body);
};

export const sendAcceptedEmail = async (email, form) => {
  const body = `Your form '${form.name}' was approved for the Publicus Marketplace by an administrator.\nIt can now be purchased any used by other users.\n${SiteUrl}/marketplace?search=${form.userId}`;
  const subject  = "Your Form was Approved For the Marketplace!";
  await sendEmail(email, subject, body);
};