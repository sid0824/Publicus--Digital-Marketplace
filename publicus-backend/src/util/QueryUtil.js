import {runQuery} from "../../libs/mysql";
import {UploadBucket} from "../../config/ConfigValues";
import {s3} from "../../libs/Aws";

const getUserNotificationSettingQuery = (surveyId) => {
  return `SELECT receiveNotifications
          FROM Surveys
          WHERE surveyId = '${surveyId}'`;
};

export const getUserNotificationSetting = async (surveyId) => {
  const response = await runQuery(getUserNotificationSettingQuery(surveyId));
  return response[0].receiveNotifications;
};

const checkAdminQuery = (userId) => {
  return `SELECT admin 
          FROM Users 
          WHERE userId = '${userId}'`;
};

export async function checkAdmin(userId) {
  const response = await runQuery(checkAdminQuery(userId));
  return response.length && response[0].admin === 1;
}

const getSurveyOwnerQuery = (surveyId) => {
  return `SELECT userId 
          FROM Surveys 
          WHERE surveyId = '${surveyId}'`;
};

export async function checkOwner(surveyId, userId) {
  const response = await runQuery(getSurveyOwnerQuery(surveyId));
  return response.length && response[0].userId === userId;
}

export const checkOwnerOrAdmin = async (surveyId, userId) => {
  const isOwner = await checkOwner(surveyId, userId);
  if (isOwner) {
    return true;
  }
  return checkAdmin(userId);
};

export const getSignedUrl = async (filePath, fileType) => {
  const s3Params = {
    Bucket: UploadBucket,
    Key: filePath,
    Expires: 500,
    ContentType: fileType,
    ACL: 'public-read'
  };
  return new Promise((resolve, reject) => {
    s3.getSignedUrl('putObject', s3Params, (err, data) => {
      if (err) {
        console.log(err);
        reject();
      }
      resolve(data);
    });
  });
};

//Generate an insert value for each tag added
export const insertTagsQuery = (surveyId, tags) => {
  let sql = "INSERT INTO Tags(surveyId, tag) VALUES ";
  for(let i = 0; i < tags.length; i++) {
    sql = sql + "( '" + surveyId + "', '" + escape(tags[i]) + "')";
    if(i < tags.length - 1) {
      sql = sql + ",";
    } else {
      sql = sql + ";";
    }
  }
  return sql;
};

export const deleteTagsBySurvey = (surveyId) => {
  return `DELETE FROM Tags WHERE surveyId = '${surveyId}'`;
};

export const getUserValuesById = (userId) => {
  return `SELECT * 
          FROM Users
          WHERE userId = '${userId}'`;
};

const checkOwnsFormQuery = (userId, surveyId) => {
  return `SELECT * 
          FROM Licenses
          WHERE userId = '${userId}' AND surveyId = '${surveyId}'`;
};

export const checkOwnsForm = async (userId, surveyId) => {
  const result = await runQuery(checkOwnsFormQuery(userId, surveyId));
  return result.length;
};
