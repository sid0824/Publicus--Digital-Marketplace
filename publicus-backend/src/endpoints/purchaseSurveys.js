import {runQuery} from "../../libs/mysql";
import {failure, success} from "../util/ResponseUtil";
import stripe from 'stripe';
import {getUserValuesById} from "../util/QueryUtil";
import {sendPurchasedEmail} from "../util/EmailUtil";
import {StripeSecretKey} from "../../config/ConfigValues";

//Generates SQL Query (Inserts licenses for each newly purchased survey)
const insertLicensesQuery = (cart, cognitoId) => {
  let sql = "INSERT INTO Licenses(surveyId, userId) VALUES ";
  for(let i = 0; i < cart.length; i++) {
      sql = sql + "( '" + cart[i] + "', '" + cognitoId + "')";
      if(i < cart.length - 1) {
          sql = sql + ",";
      } else {
          sql = sql + ";";
      }
  }
  return sql;
};

// Get survey data based on list of ids
const getSurveyPricesWithIdsQuery = (cart) => {
  let whereClause = '';
  cart.forEach((surveyId, index) => {
    if (index) {
      whereClause += `OR `;
    }
    whereClause += `surveyId = '${surveyId}'`;
  });
  return `SELECT price, userId, receiveNotifications
          FROM Surveys 
          WHERE ${whereClause}`;
};

export async function main(event, context) {
  try {
    const data = JSON.parse(event.body);
    const stripeClient = stripe(StripeSecretKey);

    // Calculate the total cost
    const surveyValues = await runQuery(getSurveyPricesWithIdsQuery(data.cart));
    let totalCost = 0;
    surveyValues.forEach((survey) => {
      totalCost += survey.price;
    });

    // Make the charge for the stripe account
    await stripeClient.charges.create({
      amount: totalCost * 100,
      currency: "AUD",
      description: "Publicus - Survey Purchase",
      source: data.tokenId
    });
    await runQuery(insertLicensesQuery(data.cart, event.requestContext.identity.cognitoIdentityId));

    //Send email to owners
    const userIds = Array.from(new Set(surveyValues.filter(value => value.receiveNotifications).map(value => value.userId)));
    await Promise.all(userIds.map(async (userId) => {
      const userData = await runQuery(getUserValuesById(userId));
      return sendPurchasedEmail(userData[0].email);
    }));

    return success();
  } catch (err) {
    console.error(err);
    return failure();
  }
}
