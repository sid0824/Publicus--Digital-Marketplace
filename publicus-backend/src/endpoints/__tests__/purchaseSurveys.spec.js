import { main } from '../purchaseSurveys';
import { runQuery } from "../../../libs/mysql";
import stripe from 'stripe';
import {sendPurchasedEmail} from "../../util/EmailUtil";

jest.mock("../../../libs/mysql");
jest.mock('stripe');
jest.mock( "../../util/EmailUtil");

const mockEvent = {
  requestContext: {
    identity: {
      cognitoIdentityId: 'MOCK_COGNITO_ID'
    },
  },
  body: JSON.stringify({
    cart: ['SURVEY_ID_1', 'SURVEY_ID_2'],
    tokenId: 'TOKEN_ID'
  }),
  pathParameters: {
    surveyId: 'MOCK_SURVEY_ID'
  }
};

const makeCharge = jest.fn();

describe('purchaseSurveys', () => {

  beforeEach(() => {
    stripe.mockReturnValue({
      charges: {
        create: makeCharge
      }
    })
  });

  afterEach(() => {
    runQuery.mockReset();
    makeCharge.mockReset();
  });

  test(`should respond with failure when an error occurs`, async () => {
    runQuery.mockImplementation(() => {
      throw new Error('ERROR')
    });
    const response = await main(mockEvent);
    expect(response).toMatchSnapshot();
  });

  test(`should purchase surveys`, async () => {
    runQuery
      .mockResolvedValueOnce([
          { price: 3, userId: 'USER_ID_1', receiveNotifications: 0 },
          { price: 4, userId: 'USER_ID_2', receiveNotifications: 1 }
        ])
      .mockResolvedValueOnce(null)
      .mockResolvedValueOnce([{ email: 'EMAIL' }]);

    const response = await main(mockEvent);
    expect(runQuery.mock.calls[0][0]).toMatchSnapshot();
    expect(runQuery.mock.calls[1][0]).toMatchSnapshot();
    expect(runQuery.mock.calls[2][0]).toMatchSnapshot();
    expect(sendPurchasedEmail.mock.calls[0][0]).toMatchSnapshot();
    expect(makeCharge.mock.calls[0][0]).toMatchSnapshot();
    expect(response).toMatchSnapshot();
  })

});