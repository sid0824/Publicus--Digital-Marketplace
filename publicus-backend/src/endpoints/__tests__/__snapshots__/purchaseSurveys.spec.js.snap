// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`purchaseSurveys should purchase surveys 1`] = `
"SELECT price, userId, receiveNotifications
          FROM Surveys 
          WHERE surveyId = 'SURVEY_ID_1'OR surveyId = 'SURVEY_ID_2'"
`;

exports[`purchaseSurveys should purchase surveys 2`] = `"INSERT INTO Licenses(surveyId, userId) VALUES ( 'SURVEY_ID_1', 'MOCK_COGNITO_ID'),( 'SURVEY_ID_2', 'MOCK_COGNITO_ID');"`;

exports[`purchaseSurveys should purchase surveys 3`] = `
"SELECT * 
          FROM Users
          WHERE userId = 'USER_ID_2'"
`;

exports[`purchaseSurveys should purchase surveys 4`] = `"EMAIL"`;

exports[`purchaseSurveys should purchase surveys 5`] = `
Object {
  "amount": 700,
  "currency": "AUD",
  "description": "Publicus - Survey Purchase",
  "source": "TOKEN_ID",
}
`;

exports[`purchaseSurveys should purchase surveys 6`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 200,
}
`;

exports[`purchaseSurveys should respond with failure when an error occurs 1`] = `
Object {
  "body": undefined,
  "headers": Object {
    "Access-Control-Allow-Credentials": true,
    "Access-Control-Allow-Origin": "*",
  },
  "statusCode": 500,
}
`;
